### DISCUSSION TITLE
Handling of "no data" "out of range" and "reserved" values

### DISCUSSION THREAD

**Discussion by aldas** [canboat/canboat#330] [Q&A] [answered]
To my understanding NMEA2000 has special values for integers to signify cases when value does not exist for that field

https://www.nmea.org/Assets/2000-explained-white-paper.pdf Page 14
> c) Every defined data format is represented by a standard data type such as Character,
Integer, Unsigned Integer, Float, or Bit Field. The data type for “Angle” is:� • uint16: 16-bit unsigned integer, range 0 to 65, 532. 65,533 = Reserved. 65,534 = Out of range. 65,535 = Data not available 

and another example 
https://www.maretron.com/support/manuals/EMS100UM_1.1.html from `EMS100 Engine Monitoring System User's Manual` `Appendix A � NMEA 2000� Interfacing`

> Note: For integer values, the most positive three values are reserved; e.g., for 8-bit unsigned integers, the values 0xFD, 0xFE, 0xFF are reserved, and for 8-bit signed integers, the values 0x7D, 0x7E, 0x7F are reserved. The most positive value (0xFF and 0x7F, respectively, for the 8-bit examples) represents Data Not Available.

Canboat has 5 different special values from 0 to -4 range. 
https://github.com/canboat/canboat/blob/fc66882597bad9066395fb30ae1f09e3dc09a76a/analyzer/analyzer.h#L70-L74

and decision if number is that special value is done here `extractNumberNotEmpty()`
https://github.com/canboat/canboat/blob/fc66882597bad9066395fb30ae1f09e3dc09a76a/analyzer/print.c#L207-L223
and 
https://github.com/canboat/canboat/blob/fc66882597bad9066395fb30ae1f09e3dc09a76a/analyzer/print.c#L237-L241

@keesverruijt there are two conditions `if (*maxValue >= 7)` and `else if (*maxValue > 1)` which decide if special value exists. 

1. Do you remember where does that `7` (`3bits`) comes?  
2. Why is reserved `DATAFIELD_RESERVED1` printed as value and not missing? Only `fieldPrintDate`, `fieldPrintLookup` prints includes reserved as special value if I search for `printEmpty` usages.

**Comment by keesverruijt** [comment, marked as answer]
The handling of special values other than ‘all ones’ (which translate to DATAFIELD_UNKNOWN) is all conjecture by me. The only sensor that I have seen producing DATAFIELD_ERROR is an old Airmar depth transducer that I don’t have anymore.

Other than some references to an ‘error value’ in the lookup lists in various Airmar docs I don’t have any reference information.

1) Well, in 1 bit you cannot have a reserved value or you end up with 1 usable value … In 2 bits there are only 4 values so having “error” available is something that I think would be excessive. I might be wrong though. There are 2 bit lookups that have 2 = Error as well as some that have 2 = some value. So this is why I assume this. I’m happy to be proven wrong, or some reference that proves me right!

2) I’m struggling to remember why the RESERVED values and these might actually be wrong.

**Comment by TwoCanPlugIn** [comment]
It is a confusing topic.

There are some values that are represented by a single bit, The value is either on/off. There is no concept of "data unavailable" or "error"
There are other values that are represented by a number of bits, again, there is no concept of "data unavailale" or "error"

Examples of these can be found in PGN 129041 AIS Aid To Navigation Report.
Field 14 Off Position Indicator is one byte, representing either On or Off Position.
Field 18 Electronic Position Fixing SystemType has the following values:
0 Undefined (default)
1 GPS
2 GLONASS
3 Combined GPS/GLONASS
4 Loran-C
5 Chayka
6 Integrated Navigation System
7 Surveyed
8 Galilleo
9-15 Reserved for Future Use

And then there are those values that are represented by bit fields that do have "data unavailable" and "error" values that are represented by the two top order values. An example of this is PGN 127233 Man Overboard Notification where field 3, MOB Emitter Status is represented by a 3 bit value with the following values:
0 - Activated
1 - Manual On-board button activated
2 - Test
3 - Not activated
4-5  - Reserved
6 - Error
7 - Unavailable 

I don't think you have any simple way of solving this problem generically. You cannot assume that for any bit field that the top two values represent "data unavailable" or "error".

Easy to do for numeric values (char, short, integer, long long and their unsigned counterparts) but difficult for values represented by bitfields other than having lookup tables for each one.

**Reply by aldas** [reply]
I think it applies only to data types that are represented as integers (canboat numbers?).  As for example canboat time and date are based on number of seconds or days since something - are those also considered into "integer" category. Lookups, I think fall into mentioned bit fields category. 

NMEA gives one example here  https://www.nmea.org/Assets/2000-explained-white-paper.pdf
![image](https://user-images.githubusercontent.com/2320301/199672689-b468f3ee-90b0-427f-ab22-57c5208125f0.png)

**Reply by aldas** [reply]
For `float` `double` types seems that magic value "data not available" is a thing. ttlappalainen NMEA2000 library https://github.com/ttlappalainen/NMEA2000/blob/579b15a1f7c8a8833295712caf539c13268d9353/src/N2kMsg.cpp#L687 seems to handle "no data" for floats. Even for `NaN` values. 

Canboat does not seems to have "magic values" for floats https://github.com/canboat/canboat/blob/af3c57d82cde1872cc662ff04a8379476a4613e3/analyzer/print.c#L352

**Reply by aldas** [reply]
This nmea publicly available document "Corrigendum TC 2000 20130720 for DSC PGN 129808"
https://www.nmea.org/Assets/20150328%20dsc%20technical%20corrigendum%20v2.database%20version%202.100.pdf

lists PGN 129808 fields.  Page 7 says that first field "DSC Format Symbol" is of type "uint8" which has range 0-252.  From that I think we can deduce that  (unsigned) integer MAX-1/2/3 are used for "special values"

**Comment by TwoCanPlugIn** [comment]
I wouldn't get too hung about what data type a value is or look for "magic" values for floats, doubles or whatever. I also think you may be misreading or misunderstanding Timo's library.

On the wire, numeric data is simply a number of bytes.  Take for example latitude. For most users, it is measured as degrees, minutes, seconds (or decimal minutes), Eg 12 degrees, 20.7 minutes. To programmers it is represented as a floating point number ie,  12.345 degrees, To provide the required level of accuracy and to be consistent in their use of SI units, NMEA scale this with a resolution of 1 e-7 and transmit it as a signed integer (4 bytes). Even though to a programmer a float is stored as 4 bytes  (sign bit, exponent & mantissa), this is very different to NMEA 2000.
 
And the simple way to detect whether the latitude is an invalid value, is not to check for a position greater than 90 degrees or less than -90 degrees, nor to check for the maximum size of a float, it is to perform the check against the signed integer value (the four bytes). So INT_MAX, INT_MAX-1 and INT_MAX-2 represent "Data not available", "Error" or "Reserved" respectively. 

Similarly for heading, 0-360 degrees, transmitted as an unsigned number using two bytes (an unsigned short) representing radians with a resolution of 1e-4., The validity check would be for USHRT_MAX, USHRT_MAX-1 and USHRT_MAX-2

This mechanism is used consistently for all numeric values. No "magic" numbers at all.

**Reply by aldas** [reply]
I am aware how numbers with fractions are transferred. I have done it a lot for Modbus.  In Modbus world this is tricky as  sometimes PLC architecture (endianess) could be different from your receiving end endianess. Even though Modbus is big-ending in the wire you still need to know in what order the sender meant for bytes to be (in what order in their memory). 

Anyway you have multiple choices how to transfer values

a) send value as integer divided with resolution and at the receiving side decode bytes into integer and multiply with resolution
b) cast your float/double/real into raw bytes and send them over the wire and receiving side needs to assemble them into float/double/real. 

Anyway in both cases you might want to have "magic"/"special" values that let receiving side know that this field does not have actually value. 

Canboat supports both cases a and b. Although (b) `float` (32bit float in this case) type is very rare - only 6 fields use it (PGN 129045 and 130321).  But it does not check for MAX-1/2 which Timo code does [ `if ( !N2kIsNA(vl) ) {`](https://github.com/ttlappalainen/NMEA2000/blob/579b15a1f7c8a8833295712caf539c13268d9353/src/N2kMsg.cpp#L690) 
 (well atleast no-data case is checked) which is used for example for [PGN 129045 user datum field 'rotation in x (deg):'](https://github.com/ttlappalainen/NMEA2000/blob/94c9bd5e1256ad8e4b3b8f7bd3f69339b2645b26/Examples/DataDisplay2/DataDisplay2.ino#L335)

same field for Canbus is defined [here](https://github.com/canboat/canboat/blob/af3c57d82cde1872cc662ff04a8379476a4613e3/analyzer/pgn.h#L3400) and printed [here](https://github.com/canboat/canboat/blob/af3c57d82cde1872cc662ff04a8379476a4613e3/analyzer/print.c#L352)
